### PR TITLE
feat: Add Tau bench environment

### LINKED ISSUE #862: Environments currently return role as part of content
https://github.com/NVIDIA-NeMo/RL/blob/0988a7dddb8c5896e28a40af1d36c63c869d2fa4/nemo_rl/environments/math_environment.py#L314-L317

This may have issues if someone uses the conversation format and apply_chat_template themselves since we may double apply the environment.

### PR DESCRIPTION
**Pull request by ashors1**
# What does this PR do ?
This PR adds a Tau bench environment to serve as an example of a multi-turn environment with an option for LLM-as-a-judge. This can be used to analyze the performance of more complicated multi-turn environments natively in NeMo-RL.

Note that we provide the option to mock user and judge models using `user_strategy='mock'` and `judge_model='mock'`.

# Issues
List issues that this PR closes ([syntax](https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)):

# Usage
* **You can potentially add a usage example below**

```python
# Add a code snippet demonstrating how to use this
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA-NeMo/RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information
* ...

### CODE REVIEW

File: nemo_rl/experience/rollouts.py [hunk 1]
```diff
@@ -55,6 +54,64 @@
...
+
+    Uses apply_chat_template to produce properly structured tokens including
+    role markers (e.g. <|im_start|>tool) and a generation prompt
+    (<|im_start|>assistant), so the model knows to respond as an assistant
+    after seeing the observation.  Raw tokenization lacks these markers and
+    causes the model to echo the observation rather than respond to it.
+
+    Falls back to raw tokenization if apply_chat_template is unavailable.
+    """
+    if not hasattr(tokenizer, "apply_chat_template"):
```
Comment (terrykong) [inline nemo_rl/experience/rollouts.py]: will this ever happen? i feel like we may make assumptions of apply_chat_template being there for most of the repo

File: nemo_rl/experience/rollouts.py [hunk 2]
```diff
@@ -55,6 +54,64 @@
...
+        add_generation_prompt=True,
+        add_special_tokens=False,
+    )
+    _without_obs = tokenizer.apply_chat_template(
+        _dummy,
+        tokenize=False,
+        add_generation_prompt=False,
+        add_special_tokens=False,
+    )
+    obs_chunk = _with_obs[len(_without_obs):]
```
Comment (terrykong) [inline nemo_rl/experience/rollouts.py]: maybe we should do an assert above here like:
```suggestion
    assert _with_obs[len(_without_obs):] == _without_obs, "expected them to match...."
    obs_chunk = _with_obs[len(_without_obs):]
```
just to make sure the templating is correct and nothing weird is happening

File: pyproject.toml [hunk 3]
```diff
@@ -550,6 +525,22 @@ requires-dist = [
   "requests>=2.32.5,<3.0.0",
 ]
 
+# Override tau-bench to remove google-generativeai dependency: it introduces irreconcilable
+# protobuf version conflicts. NeMo-RL only uses tau-bench's environments, not its Google model
+# integrations; litellm handles all model calls via OpenAI-compatible endpoints.
+[[tool.uv.dependency-metadata]]
```
Comment (terrykong) [inline pyproject.toml:617]: out of curiosity, if we only need to forbid google-generativeai, would it have worked to just do something like this https://github.com/astral-sh/uv/issues/9174#issuecomment-2482710850 so we dont' have to maintain dependency-meta for this case

File: nemo_rl/experience/rollouts.py [hunk 4]
```diff
@@ -54,6 +54,64 @@
 TokenizerType = PreTrainedTokenizerBase
 
 
+def _tokenize_env_observation(
```
Comment (terrykong) [inline nemo_rl/experience/rollouts.py:294]: [`rollouts.py:57`](https://github.com/NVIDIA-NeMo/RL/blob/c0c27ec13a1b878debb0a0b403a0b36c2840f1a1/nemo_rl/experience/rollouts.py#L57) — **Critical regression risk**

This function passes `obs_role` directly into `apply_chat_template`, but all existing multi-turn environments emit `role: "environment"`:
- [`math_environment.py:486`](https://github.com/NVIDIA-NeMo/RL/blob/c0c27ec13a1b878debb0a0b403a0b36c2840f1a1/nemo_rl/environments/math_environment.py#L486)
- [`code_environment.py:145`](https://github.com/NVIDIA-NeMo/RL/blob/c0c27ec13a1b878debb0a0b403a0b36c2840f1a1/nemo_rl/environments/code_environment.py#L145)
- [`sliding_puzzle.py:296`](https://github.com/NVIDIA-NeMo/RL/blob/c0c27ec13a1b878debb0a0b403a0b36c2840f1a1/nemo_rl/environments/games/sliding_puzzle.py#L296)
- [`vlm_environment.py:195`](https://github.com/NVIDIA-NeMo/RL/blob/c0c27ec13a1b878debb0a0b403a0b36c2840f1a1/nemo_rl/environments/vlm_environment.py#L195)

`"environment"` is not a recognized chat-template role. Reproduced with Qwen2.5: the template silently drops the message, so `_with_obs[len(_without_obs):]` yields only the assistant generation prompt — the observation content is never tokenized.

Consider either gating this behind a per-env opt-in, or mapping `"environment"` to a known role before calling `apply_chat_template`. Could you also share convergence results on an existing env (e.g. math or sliding_puzzle) to confirm no regression?

Also: the fallback at [line 72](https://github.com/NVIDIA-NeMo/RL/blob/c0c27ec13a1b878debb0a0b403a0b36c2840f1a1/nemo_rl/experience/rollouts.py#L72) (`if not hasattr(tokenizer, "apply_chat_template")`) is dead code — the method exists on all HF tokenizers. If the intent is to handle tokenizers without a chat template string, check `getattr(tokenizer, "chat_template", None) is None` instead.

This function also has zero test coverage — suggest adding tests with a real HF tokenizer covering `obs_role="tool"`, `obs_role="user"`, empty content, and the fallback path.

File: nemo_rl/experience/rollouts.py [hunk 5]
```diff
@@ -54,6 +54,64 @@
...
+    #      non-final message vs a final message).  If violated, _without_obs
+    #      ends without a trailing "\n" while the same message inside _with_obs
+    #      does, making the string-length cutpoint off by one character.
+    #   2. Re-tokenizing the extracted string chunk gives the same token IDs as
+    #      those tokens would have in the full sequence context (i.e. the
+    #      tokenizer is not context-dependent at the observation boundary).
+    #      This holds for all models using special tokens for role markers
+    #      (e.g. <|im_start|>, <|eot_id|>), which cannot be merged with
+    #      adjacent text by BPE.
+    _dummy = [{"role": "user", "content": "x"}]
```
Comment (terrykong) [inline nemo_rl/experience/rollouts.py]: [`rollouts.py:96`](https://github.com/NVIDIA-NeMo/RL/blob/c0c27ec13a1b878debb0a0b403a0b36c2840f1a1/nemo_rl/experience/rollouts.py#L96)

The dummy-message slicing assumes `_without_obs` is always a strict prefix of `_with_obs`. As the comment acknowledges, templates with `loop.last` logic (e.g. Llama-3) may change trailing separators for the last message, breaking this assumption.

Consider adding a runtime check like `assert _with_obs.startswith(_without_obs)`, or verifying this works across Qwen/Llama/Mistral templates.

File: tests/unit/environments/test_tau_bench_environment.py [hunk 6]
```diff
@@ -0,0 +1,441 @@
...
+# Helpers: instantiate the underlying (non-Ray) classes for unit testing
+# ---------------------------------------------------------------------------
+
+def _make_worker(
+    judge_model=None,
+    judge_base_url=None,
+    judge_api_key="dummy",
+    max_steps=30,
+):
+    """Return a bare TauBenchWorker instance without spawning a Ray actor."""
```
Comment (terrykong) [inline tests/unit/environments/test_tau_bench_environment.py:89]: [`test_tau_bench_environment.py:74-92`](https://github.com/NVIDIA-NeMo/RL/blob/c0c27ec13a1b878debb0a0b403a0b36c2840f1a1/tests/unit/environments/test_tau_bench_environment.py#L74) — **Broken test fixture**

`_make_worker()` never sets `_mock_mode` or `_mock_latency_s`. When `_call_judge` runs `if self._mock_mode:` ([line 266](https://github.com/NVIDIA-NeMo/RL/blob/c0c27ec13a1b878debb0a0b403a0b36c2840f1a1/nemo_rl/environments/tau_bench_environment.py#L266)), it raises `AttributeError`.

Additionally, all `TestCallJudge` tests mock `requests.post` but the implementation uses [`litellm.completion`](https://github.com/NVIDIA-NeMo/RL/blob/c0c27ec13a1b878debb0a0b403a0b36c2840f1a1/nemo_rl/environments/tau_bench_environment.py#L281). The mocks are never hit — tests either fail or pass for the wrong reason (the broad `except Exception` returns 0.0).

Suggested fix: add `worker._mock_mode = False` and `worker._mock_latency_s = 0.0` to `_make_worker()`, then replace `mock.patch("requests.post", ...)` with `mock.patch("litellm.completion", ...)` and build mocks matching litellm's return type (`resp.choices[0].message.content`).

File: nemo_rl/environments/tau_bench_environment.py [hunk 7]
```diff
@@ -0,0 +1,564 @@
...
+    ) -> None:
+        self._env_name = env_name
+        self._task_split = task_split
+        self._user_strategy = user_strategy
+        self._user_model = user_model
+        self._max_steps = max_steps
+        self._judge_model = judge_model
+        self._judge_base_url = judge_base_url
+        self._judge_api_key = judge_api_key
+        self._mock_mode = user_strategy == "mock"
```
Comment (terrykong) [inline nemo_rl/environments/tau_bench_environment.py]: [`tau_bench_environment.py:140`](https://github.com/NVIDIA-NeMo/RL/blob/c0c27ec13a1b878debb0a0b403a0b36c2840f1a1/nemo_rl/environments/tau_bench_environment.py#L140)

`_mock_mode` is gated only on `user_strategy == "mock"`, not on `judge_model`. The exemplar config sets `judge_model: "mock"`, but if a user changes `user_strategy: "llm"` while leaving `judge_model: "mock"`, `litellm.completion(model="mock")` will error and silently return 0.0 via the `except Exception` handler.

Consider gating mock mode for the judge independently, or documenting that the two flags must change together.

File: nemo_rl/environments/tau_bench_environment.py [hunk 8]
```diff
@@ -0,0 +1,564 @@
...
+                    "tau_reward": None,
+                    "judge_score": None,
+                })
+                continue
+
+            action = self._parse_action(agent_text)
+            # Snapshot before step(): when done=True, step() calls calculate_reward()
+            # internally which replays gold actions via self.step(), appending them to
+            # tau_env.actions and corrupting any snapshot taken afterwards.
+            actions_snapshot = list(tau_env.actions) + [action]
```
Comment (terrykong) [inline nemo_rl/environments/tau_bench_environment.py]: [`tau_bench_environment.py:362`](https://github.com/NVIDIA-NeMo/RL/blob/c0c27ec13a1b878debb0a0b403a0b36c2840f1a1/nemo_rl/environments/tau_bench_environment.py#L362)

`actions_snapshot` is computed but never read. The comment explains the intent (snapshot before `step()` corrupts `tau_env.actions`), but no downstream code consumes it. Could you confirm whether this is leftover from a removed code path, or if there is missing follow-up logic?

File: nemo_rl/environments/tau_bench_environment.py [hunk 9]
```diff
@@ -0,0 +1,564 @@
...
+    ``{"task_index": <int>, "episode_id": null, "step_count": 0}``.  The initial
+    ``message_log`` should be constructed by calling ``tau_bench.envs.get_env`` and
+    ``env.reset(task_index=N)`` during data preparation to ensure prompt/env
+    consistency.
+    """
+
+    def __init__(self, cfg: TauBenchEnvConfig) -> None:
+        self.cfg = cfg
+        self._num_workers = cfg["num_workers"]
+        self._judge_weight = float(cfg.get("judge_weight", 0.0))
```
Comment (terrykong) [inline nemo_rl/environments/tau_bench_environment.py]: [`tau_bench_environment.py:451-468`](https://github.com/NVIDIA-NeMo/RL/blob/c0c27ec13a1b878debb0a0b403a0b36c2840f1a1/nemo_rl/environments/tau_bench_environment.py#L451)

Six `cfg.get(..., <non-None default>)` calls here violate the [config-conventions guideline](https://github.com/NVIDIA-NeMo/RL/blob/c0c27ec13a1b878debb0a0b403a0b36c2840f1a1/skills/config-conventions/SKILL.md). The Python defaults also diverge from the YAML values — e.g. `max_steps` defaults to 30 here but 50 in the YAML, `mock_stop_prob` defaults to 0.2 here but 0.4 in the YAML.

Consider making these required in YAML (they are already present in `grpo_tau_bench.yaml`) and using `cfg[key]` or `cfg.get(key)` without a fallback.

File: examples/configs/grpo_tau_bench.yaml [hunk 10]
```diff
@@ -0,0 +1,113 @@
+# GRPO training on tau-bench: multi-turn customer service tasks with LLM-as-judge rewards.
+#
+# The agent must complete realistic retail or airline service tasks by calling domain
+# tools across multiple turns.  Rewards blend tau-bench's deterministic task-completion
+# check with an optional LLM judge score for policy compliance.
+#
+# Data preparation: run scripts/prepare_tau_bench_data.py to generate the JSONL files
```
Comment (terrykong) [inline examples/configs/grpo_tau_bench.yaml]: [`grpo_tau_bench.yaml:7-10`](https://github.com/NVIDIA-NeMo/RL/blob/c0c27ec13a1b878debb0a0b403a0b36c2840f1a1/examples/configs/grpo_tau_bench.yaml#L7)

This comment references `scripts/prepare_tau_bench_data.py`, but no such script exists. `TauBenchDataset` builds records at runtime via `tau_bench.envs.get_env` — no offline data prep is needed. The [`tau_bench_data_processor` docstring](https://github.com/NVIDIA-NeMo/RL/blob/c0c27ec13a1b878debb0a0b403a0b36c2840f1a1/nemo_rl/data/processors.py#L719) also references `examples/prepare_tau_bench_data.py`. Both should be updated to reflect the actual data flow.

File: pyproject.toml [hunk 11]
```diff
@@ -164,6 +165,7 @@ megatron-core = { workspace = true }
 nemo-automodel = { path = "3rdparty/Automodel-workspace/Automodel", editable = true }
 megatron-bridge = { path = "3rdparty/Megatron-Bridge-workspace", editable = true }
 nemo_gym = { workspace = true }
+tau-bench = { git = "https://github.com/sierra-research/tau-bench" }
```
Comment (terrykong) [inline pyproject.toml]: [`pyproject.toml:168`](https://github.com/NVIDIA-NeMo/RL/blob/c0c27ec13a1b878debb0a0b403a0b36c2840f1a1/pyproject.toml#L168)

This git dependency has no `rev` or `tag` pin, unlike all other git dependencies in this file (e.g. [`nemo_run` at line 169](https://github.com/NVIDIA-NeMo/RL/blob/c0c27ec13a1b878debb0a0b403a0b36c2840f1a1/pyproject.toml#L169)). Builds on different days may pick up different tau-bench commits, affecting reproducibility. Consider pinning to a specific commit.

File: nemo_rl/data/processors.py [hunk 12]
```diff
@@ -707,6 +707,62 @@ def nemo_gym_data_processor(
     return output
 
 
+def tau_bench_data_processor(
```
Comment (terrykong) [inline nemo_rl/data/processors.py:806]: [`processors.py:710`](https://github.com/NVIDIA-NeMo/RL/blob/c0c27ec13a1b878debb0a0b403a0b36c2840f1a1/nemo_rl/data/processors.py#L710)

`tau_bench_data_processor` is registered in `PROCESSOR_REGISTRY` but has no tests. Consider adding basic tests: returns valid `DatumSpec`, passes through `extra_env_info`, and handles the truncation path when `len(token_ids) >= max_seq_length`.

File: nemo_rl/experience/rollouts.py (hunk 4 above)
Comment (terrykong) [reply nemo_rl/experience/rollouts.py:294]: this makes sense to maybe default, but what should we do about the existing environments. i'm guessing the multi-turn sliding puzzle and any of the other environments will have slightly different trajectories.

actually what happens for openmathinstruct in this case?

**Review by terrykong — COMMENTED** [review]
## Review Summary

Nice work adding the tau-bench multi-turn environment — the architecture is clean and the mock mode is a thoughtful addition for testing without LLM credentials. Two critical issues need attention before merge:

1. **`_tokenize_env_observation` silently breaks all existing multi-turn environments** — the new function passes `role="environment"` to `apply_chat_template`, which silently drops the message on Qwen/Llama templates. Math, code, sliding_puzzle, vlm, and reward_model environments would lose all observation feedback in rollouts.

2. **`TestCallJudge` is entirely broken** — fixture missing `_mock_mode` attribute, and all tests mock `requests.post` instead of `litellm.completion`.

### Additional notes (not tied to specific diff lines)

- **Documentation**: This is a new feature (`feat:`) but there is no user-facing documentation. `docs/guides/environments.md` does not mention tau-bench, and there is no guide analogous to `docs/guides/grpo-sliding-puzzle.md`. Consider adding a section.

- **Design note**: The pre-step mechanism wastes one generation turn per episode (~2% overhead with `max_steps=50`). The tradeoff is documented in the code and reasonable, but worth noting.

- **Global litellm monkey-patch**: Mock mode at [tau_bench_environment.py:149](https://github.com/NVIDIA-NeMo/RL/blob/c0c27ec13a1b878debb0a0b403a0b36c2840f1a1/nemo_rl/environments/tau_bench_environment.py#L149) patches `litellm.completion` globally at the module level. Any other code in the same worker process that uses `litellm.completion` would also get the mock. Consider a more targeted patch.

- **Env var clobbering**: `os.environ["OPENAI_API_KEY"] = user_api_key` at [tau_bench_environment.py:188](https://github.com/NVIDIA-NeMo/RL/blob/c0c27ec13a1b878debb0a0b403a0b36c2840f1a1/nemo_rl/environments/tau_bench_environment.py#L188) unconditionally overwrites any existing key (defaults to `"dummy"` from config fallback). Guard with `if user_api_key:`.

_Generated by Claude Code_

File: nemo_rl/experience/rollouts.py (hunk 1 above)
Comment (ashors1) [reply nemo_rl/experience/rollouts.py]: Good point. I've changed the logic in this section a bit, but I've also added an explicit assertion just to be sure that `chat_template` is present when we expect it to be

File: nemo_rl/experience/rollouts.py (hunk 4 above)
Comment (ashors1) [reply nemo_rl/experience/rollouts.py:294]: Good catch. To make sure we don't regress, I added a `obs_use_chat_template` property that determines whether we should apply the chat template to environment observations. This defaults to False, so it shouldn't break any existing environments

File: tests/test_suites/disabled.txt [hunk 13]
```diff
@@ -7,3 +7,6 @@ tests/test_suites/llm/grpo-qwen3.5-35ba3b-2n8g-automodel-ep16.sh
 
 # Disable this due to know vLLM bug with Qwen3.5: https://github.com/vllm-project/vllm/issues/37856
 tests/test_suites/llm/grpo-qwen3.5-397ba17b-32n8g-megatron.v2.sh
+
+# Requires external user-simulation and judge servers (USER_SERVER_HOST:8100); not available in CI.
+tests/test_suites/llm/grpo_tau_bench_local.sh
```
Comment (terrykong) [inline tests/test_suites/disabled.txt]: [`disabled.txt:12`](https://github.com/NVIDIA-NeMo/RL/blob/968b1ff0256b4d5703b84baf6db9fa188799ac3c/tests/test_suites/disabled.txt#L12)

**CI-breaking.** This disables `grpo_tau_bench_local.sh`, but the committed script and recipe are named `grpo-qwen2.5-7b-instruct-16n8g-tau-bench-local` — the real name is in no suite list. [`tests/unit/test_recipes_and_test_suites.py`](https://github.com/NVIDIA-NeMo/RL/blob/968b1ff0256b4d5703b84baf6db9fa188799ac3c/tests/unit/test_recipes_and_test_suites.py#L213) asserts exact set-equality between on-disk `.sh` files, recipe YAMLs, and suite entries, so `test_all_test_scripts_accounted_for_in_test_suites`, `test_all_recipe_yamls_accounted_for_in_test_suites`, and `test_all_tests_can_find_config_if_dryrun` all fail (reproduced locally). This is likely why the merge is blocked.

**Action:** point this entry at the actual filename:
```suggestion
tests/test_suites/llm/grpo-qwen2.5-7b-instruct-16n8g-tau-bench-local.sh
```

File: nemo_rl/environments/tau_bench_environment.py [hunk 14]
```diff
@@ -0,0 +1,706 @@
...
+                    {"role": "user", "content": user_prompt},
+                ],
+                temperature=0.0,
+                timeout=60,
+                api_base=self._judge_base_url,
+                api_key=self._judge_api_key,
+            )
+            content = resp.choices[0].message.content
+            return float(json.loads(content).get("score", 0.0))
+        except Exception as e:
```
Comment (terrykong) [inline nemo_rl/environments/tau_bench_environment.py]: [`tau_bench_environment.py:336`](https://github.com/NVIDIA-NeMo/RL/blob/968b1ff0256b4d5703b84baf6db9fa188799ac3c/nemo_rl/environments/tau_bench_environment.py#L336)

`_call_judge` collapses every error to `0.0` with **no retry**, which lumps two very different failures together:

- **Transient / infra** (timeout, connection refused, rate limit, 5xx, judge server not up yet) — not the agent's fault and not a real score of 0. With `judge_weight: 0.3`, a brief judge outage silently scores affected episodes `0.7*tau_reward` and writes a fake `0.0` into `judge_score` → pollutes `tau_bench/mean_judge_score`, with nothing in the metrics to reveal it.
- **Parse / content** (`json.JSONDecodeError`, missing `score`) — the judge replied but the output is unusable. Since the judge runs at `temperature=0.0`, a retry reproduces the same output, so retry won't help here; but these should still be *counted* so an always-mis-formatting judge (all-zeros signal) is visible.

There's already precedent for this disambiguation in this same file: the `tau_env.step` retry loop branches on the litellm exception class, treating [`litellm.BadRequestError` as non-retryable](https://github.com/NVIDIA-NeMo/RL/blob/968b1ff0256b4d5703b84baf6db9fa188799ac3c/nemo_rl/environments/tau_bench_environment.py#L417) and retrying everything else 5×. The judge path just doesn't mirror it.

**Action:** catch by exception class (litellm re-exports the OpenAI-style typed exceptions, so no need to parse HTTP status codes):

```python
for attempt in range(5):
    try:
        resp = litellm.completion(model=self._judge_model, ..., timeout=60)
        return float(json.loads(resp.choices[0].message.content)["score"])
    except (litellm.Timeout, litellm.APIConnectionError, litellm.RateLimitError,
            litellm.InternalServerError, litellm.ServiceUnavailableError) as e:
        last_exc = e                       # transient -> back off and retry
        time.sleep(2**attempt + random.uniform(0, 1))
    except (json.JSONDecodeError, KeyError, ValueError):
        self._judge_parse_failures += 1    # deterministic at temp=0; count, don't retry
        return None
return None  # retries exhausted -> a real outage, NOT a real 0
```

And the part that matters most for the training signal: on `None`, **don't blend `judge_weight*0`** — fall back to `reward = tau_reward` for that sample (score it as if the judge were disabled this episode) and surface a `judge_failures` count in the metrics, so "judge was down" is never indistinguishable from "agent scored 0." (If a hard `raise` is too aggressive for an example env, the minimum bar is transient-retry + a logged failure counter.)

File: nemo_rl/environments/tau_bench_environment.py [hunk 15]
```diff
@@ -0,0 +1,706 @@
...
+        self._workers = [
+            TauBenchWorker.options(
+                runtime_env={"py_executable": PY_EXECUTABLES.TAU_BENCH}
+            ).remote(
+                env_name=cfg["env_name"],
+                task_split=cfg["task_split"],
+                user_strategy=cfg["user_strategy"],
+                user_model=cfg["user_model"],
+                user_base_url=cfg.get("user_base_url"),
+                user_api_key=cfg.get("user_api_key") or "dummy",
```
Comment (terrykong) [inline nemo_rl/environments/tau_bench_environment.py]: [`tau_bench_environment.py:597`](https://github.com/NVIDIA-NeMo/RL/blob/968b1ff0256b4d5703b84baf6db9fa188799ac3c/nemo_rl/environments/tau_bench_environment.py#L597)

**What to change:** this line. The worker added an `if user_api_key:` guard ([L216](https://github.com/NVIDIA-NeMo/RL/blob/968b1ff0256b4d5703b84baf6db9fa188799ac3c/nemo_rl/environments/tau_bench_environment.py#L216)) to avoid clobbering an ambient `OPENAI_API_KEY`, but this call site passes `cfg.get("user_api_key") or "dummy"` — always truthy — so the guard always fires, the `elif "OPENAI_API_KEY" not in os.environ` fallback ([L218](https://github.com/NVIDIA-NeMo/RL/blob/968b1ff0256b4d5703b84baf6db9fa188799ac3c/nemo_rl/environments/tau_bench_environment.py#L218)) is dead code, and a user who omits `user_api_key` while relying on a real ambient `OPENAI_API_KEY` (real-OpenAI user simulator) has it overwritten with `"dummy"`. (It's also a call-site default, which [config-conventions](https://github.com/NVIDIA-NeMo/RL/blob/968b1ff0256b4d5703b84baf6db9fa188799ac3c/.claude/skills/config-conventions/SKILL.md) forbids — the `"dummy"` default belongs in the worker's env fallback, not here.)

**Action:** drop the `or "dummy"` so the worker receives `None` when the key is unset and its existing `elif "OPENAI_API_KEY" not in os.environ` branch supplies `"dummy"` only when there's no ambient key:

```suggestion
                user_api_key=cfg.get("user_api_key"),
```

(You may also want to loosen the worker's `user_api_key: str` annotation to `Optional[str]` at [L129](https://github.com/NVIDIA-NeMo/RL/blob/968b1ff0256b4d5703b84baf6db9fa188799ac3c/nemo_rl/environments/tau_bench_environment.py#L129).)

_Note:_ `judge_api_key` on [L601](https://github.com/NVIDIA-NeMo/RL/blob/968b1ff0256b4d5703b84baf6db9fa188799ac3c/nemo_rl/environments/tau_bench_environment.py#L601) does **not** have the clobber problem — it's passed straight to `litellm.completion(api_key=...)` ([L332](https://github.com/NVIDIA-NeMo/RL/blob/968b1ff0256b4d5703b84baf6db9fa188799ac3c/nemo_rl/environments/tau_bench_environment.py#L332)) and never written to `os.environ`. Its `or "dummy"` is only the milder call-site-default smell, so cleaning it up is optional.

File: nemo_rl/environments/tau_bench_environment.py [hunk 16]
```diff
@@ -0,0 +1,706 @@
...
+            the user simulator and judge return random outputs after an optional sleep,
+            and user_base_url / judge_base_url are ignored.
+        user_model: LiteLLM model string for the user simulator (e.g. "openai/gpt-4o-mini").
+            Ignored when user_strategy is "mock".
+        user_base_url: Optional base URL override for the user model's OpenAI-compatible API.
+            Ignored when user_strategy is "mock".
+        user_api_key: API key for the user model server. Ignored when user_strategy is "mock".
+        max_steps: Maximum tool-call steps per episode before forced termination.
+        judge_model: LiteLLM model string for LLM-as-judge scoring (None disables judging).
+            Ignored when user_strategy is "mock".
```
Comment (terrykong) [inline nemo_rl/environments/tau_bench_environment.py]: [`tau_bench_environment.py:65-69`](https://github.com/NVIDIA-NeMo/RL/blob/968b1ff0256b4d5703b84baf6db9fa188799ac3c/nemo_rl/environments/tau_bench_environment.py#L65)

These docstrings say `judge_model`/`judge_base_url`/`judge_api_key` are "Ignored when `user_strategy` is 'mock'", but the judge is mocked independently — gated on `judge_model == "mock"` ([L155](https://github.com/NVIDIA-NeMo/RL/blob/968b1ff0256b4d5703b84baf6db9fa188799ac3c/nemo_rl/environments/tau_bench_environment.py#L155)), not on `user_strategy`. So `user_strategy="mock"` + a real `judge_model` still issues real judge calls.

**Action:** update lines 66/68/69 to "Ignored when `judge_model` is 'mock'."

File: nemo_rl/experience/rollouts.py [hunk 17]
```diff
@@ -230,6 +230,88 @@ def _apply_effort_shaping(
     )
 
 
+def _tokenize_env_observation(
```
Comment (terrykong) [inline nemo_rl/experience/rollouts.py:294]: [`rollouts.py:233`](https://github.com/NVIDIA-NeMo/RL/blob/968b1ff0256b4d5703b84baf6db9fa188799ac3c/nemo_rl/experience/rollouts.py#L233)

`_tokenize_env_observation(use_chat_template=True)` — the dummy-primed token-slicing routine central to tau-bench correctness — has no test coverage (grep-confirmed). Its own docstring lists real failure modes (`loop.last` separator drift, context-dependent BPE), so it's worth locking down.

**Action:** add these tests to `tests/unit/experience/test_rollouts.py` (verified passing locally against Qwen2.5-1.5B-Instruct via the existing `rollout_tokenizer` fixture):

```python
from nemo_rl.experience.rollouts import _tokenize_env_observation


def test_tokenize_env_observation_raw(rollout_tokenizer):
    out = _tokenize_env_observation(
        rollout_tokenizer, "tool", "hello world", use_chat_template=False
    )
    expected = rollout_tokenizer(
        "hello world", return_tensors="pt", add_special_tokens=False
    ).input_ids[0]
    assert torch.equal(out, expected)


def test_tokenize_env_observation_tool_chat_template(rollout_tokenizer):
    out = _tokenize_env_observation(
        rollout_tokenizer, "tool", "ORDER_STATUS: shipped", use_chat_template=True
    )
    decoded = rollout_tokenizer.decode(out)
    assert "ORDER_STATUS: shipped" in decoded
    assert decoded.startswith("<|im_end|>")
    assert decoded.rstrip().endswith("<|im_start|>assistant")
    assert "<tool_response>" in decoded  # Qwen wraps tool results in a user turn


def test_tokenize_env_observation_user_chat_template(rollout_tokenizer):
    out = _tokenize_env_observation(
        rollout_tokenizer, "user", "Can you help me?", use_chat_template=True
    )
    decoded = rollout_tokenizer.decode(out)
    assert decoded.startswith("<|im_end|>")
    assert "<|im_start|>user" in decoded
    assert "Can you help me?" in decoded
    assert decoded.rstrip().endswith("<|im_start|>assistant")


def test_tokenize_env_observation_excludes_dummy_priming(rollout_tokenizer):
    # The dummy "x" priming message must not leak into the returned tokens.
    out = _tokenize_env_observation(
        rollout_tokenizer, "user", "unique_marker_zzz", use_chat_template=True
    )
    decoded = rollout_tokenizer.decode(out)
    assert "unique_marker_zzz" in decoded
    assert "<|im_start|>user\nx<|im_end|>" not in decoded
```

File: nemo_rl/environments/tau_bench_environment.py [hunk 18]
```diff
@@ -0,0 +1,706 @@
...
+
+        action = self._parse_action(agent_text)
+        last_exc: Exception = RuntimeError("unreachable")
+        env_response = None
+        for attempt in range(5):
+            try:
+                if self._mock_user:
+                    from unittest.mock import patch
+
+                    with patch("litellm.completion", self._mock_completion):
```
Comment (terrykong) [inline nemo_rl/environments/tau_bench_environment.py]: [`tau_bench_environment.py:406`](https://github.com/NVIDIA-NeMo/RL/blob/968b1ff0256b4d5703b84baf6db9fa188799ac3c/nemo_rl/environments/tau_bench_environment.py#L406)

Non-blocking (only reachable with `user_strategy="mock"` + `max_concurrent_api_requests>1`, which no shipped recipe uses). [`_execute_one`](https://github.com/NVIDIA-NeMo/RL/blob/968b1ff0256b4d5703b84baf6db9fa188799ac3c/nemo_rl/environments/tau_bench_environment.py#L340) runs concurrently in a `ThreadPoolExecutor` (default 8), and each thread does `with patch("litellm.completion", ...)`. `unittest.mock.patch` mutates a process-global and isn't thread-safe, so overlapping threads leak the patch — a mock-mode step can call the real `litellm.completion`, and the global can be left permanently patched (reproduced: 8 leak events at 8 workers). The inline comment's "other litellm users are not affected" doesn't hold under concurrency.

**Action:** avoid patching the global under concurrency — inject the mock into tau-bench's user simulator, guard the patched region with a lock, or force `max_concurrent=1` in mock mode.

File: nemo_rl/environments/tau_bench_environment.py [hunk 19]
```diff
@@ -0,0 +1,706 @@
...
+        tau_rewards: List[float] = []
+        judge_scores: List[float] = []
+        for meta in batch.get("extra_env_info", []):
+            if meta and meta.get("tau_reward") is not None:
+                tau_rewards.append(float(meta["tau_reward"]))
+            if meta and meta.get("judge_score") is not None:
+                judge_scores.append(float(meta["judge_score"]))
+
+        metrics: Dict[str, Any] = {
+            "tau_bench/task_completion_rate": rewards.mean().item(),
```
Comment (terrykong) [inline nemo_rl/environments/tau_bench_environment.py]: [`tau_bench_environment.py:693`](https://github.com/NVIDIA-NeMo/RL/blob/968b1ff0256b4d5703b84baf6db9fa188799ac3c/nemo_rl/environments/tau_bench_environment.py#L693)

`tau_bench/task_completion_rate` is computed from the blended `rewards*is_end`, so when `judge_weight>0` it reports the tau+judge blend, not the true completion rate (`mean_tau_reward` already reports the pure value).

**Action:** rename this to `mean_reward`, or compute it from the `tau_reward` metadata so the label matches its value.

**Review by terrykong — COMMENTED** [review]
## Re-review — tau-bench environment

Thanks for the thorough follow-ups — the entire prior review is addressed: `obs_use_chat_template()` cleanly gates the chat-template path (defaults False, so existing environments are untouched), the `TestCallJudge` fixture/mocks are fixed (31 tests pass locally), and the API-key / mock-gating / dead-code / dependency-pin / docs items are all resolved. The `get_idx_grouping` refactor is also correct — it fixes the multi-turn shared-prompt grouping case.

One blocker remains, plus a few low-severity suggestions inline:

- **Blocker:** `tests/test_suites/disabled.txt` points at a stale filename → 3 unit tests in `test_recipes_and_test_suites.py` fail (reproduced locally). Likely the cause of the blocked merge.
- Low-severity: judge error handling, ambient `OPENAI_API_KEY` clobber, a docstring fix, an uncovered token-slicing routine (with a ready-to-drop test), a mock-mode thread-safety note, and a metric label.

All 121 unit tests across the 4 changed test files pass, and `pre-commit` is clean on the PR's files.

_Generated by Claude Code_
